docs/Plugins-Guide.md

```diff
@@ -0,0 +1,232 @@
+  reply.send(req)
+})
+
+fastify.get('/plugin2', (req, reply) => {
+  reply.send(req)
+})
+```
+Now for every request you will run your utility, it is obvious that you can register as many hooks as you need.  
+It can happen that you want that a hook must be executed just for a subset of routes, how can you do that?  Yep, encapsulation!
+
+```js
+fastify.register((instance, opts, next) => {
+  instance.decorate('util', (req, key, value) => { req.key = value })
+
+  instance.addHook('preHandler', (req, reply, done) => {
+    instance.util(req, 'timestamp', new Date())
+    done()
+  })
+
+  instance.get('/plugin1', (req, reply) => {
+    reply.send(req)
+  })
+
+  next()
+})
+
+fastify.get('/plugin2', (req, reply) => {
+  reply.send(req)
+})
+```
+Now your hook will run just for the first route!
+
+As you probably noticed at this time, `request` and `reply` are not the standard Nodejs *request* and *response* objects, but Fastify's objects.  
+Let's say that you are arriving from a framework like Express or Restify, and you already have some Middleware that do exactly what you need, and you don't want to redo al the work.
+
+<a name="middlewares"></a>
+## Middlewares
+Fastify [supports](https://github.com/fastify/fastify/blob/master/docs/Middlewares.md) out of the box Express/Restify middlewares this means that you can just drop-in your old code and it will work! (faster, by the way)  
+How we can do that? Checkout our middlewares engine, [middie](https://github.com/fastify/middie).
+```js
+const yourMiddleware = require('your-middleware')
+fastify.use(yourMiddleware)
+```
+
+<a name="distribution"></a>
+## How handle encapsulation and distribution
+Perfect, now you know (almost) all the tools that you can use to extend Fastify. But probably there is something you noted when trying out your code.  
+How can you distribute your code?
+
+The easy way, is to distribute an utility and tell to your users to decorate their Fastify instance with you function. But what happen if your utility has na asynchronous bootstrapping? A database connection for example.
+`decorate` is a synchronous api, so it cannot solve this kind of issue.  
+You already know what your are going to read, I'm pretty sure about that.
+
+Exactly, `register`!!
+
+*Wait, what? Don't you told me that `register` creates and encapsulation and what I create iside there will not be available outside?*  
+Yes, I told that. But what I didn't told you, is that you can tell to Fastify to avoid this behavior, with the [`fastify-plugin`](https://github.com/fastify/fastify-plugin) module.
+```js
+const fp = require('fastify-plugin')
+const dbClient = require('db-client')
+
+function dbPlugin (fastify, opts, next) {
+  dbClient.connect(opts.url, (err, conn) => {
+    fastify.decorate('db', conn)
+    next()
+  })
+}
+
+module.exports = fp(dbPlugin)
+```
+That was easy right? And you can also tell to `fastify-plugin` to check the installed version of Fastify, in case of you need a specific api.
```

Comment:
Calling something `'easy'` might put people off.

You should also add a reference in https://github.com/fastify/fastify/pull/143/files#diff-4edceccedc5b5f87a59494e099c06593R112 to this paragraph. This is the preferred way to still using plugins and avoid contexts.